### PR TITLE
[Feature] Support decode with DP in pytorch

### PR DESCRIPTION
**Pull request by wangruohui**
This is based on refactored `init_model` in #178. 

## Motivation

Support decode ( compute probabilities of input prompt ) with data parallel in pytorch module.

## Modification

New feature. 

## BC-breaking (Optional)

No. 

## Use cases (Optional)

If this PR introduces a new feature, it is better to list some use cases here, and update the documentation.

## Checklist

1. Pre-commit or other linting tools are used to fix the potential lint issues.
2. The modification is covered by complete unit tests. If not, please add more unit tests to ensure the correctness.
3. If the modification has a dependency on downstream projects of a newer version, this PR should be tested with all supported versions of downstream projects.
4. The documentation has been modified accordingly, like docstring or example tutorials.

### CODE REVIEW

**Comment by varuy322** [comment]
please add pynvml or deepspeed in requirement.txt.
It would be nice to add an example, such as 8XA100  for inference.
thanks a ton~

File: lmdeploy/pytorch/decode.py [hunk 1]
```diff
@@ -0,0 +1,327 @@
...
+
+    def decode(self,
+               prompt_token_ids: List[List[int]],
+               sort=True,
+               max_bs: int = 1024,
+               pad=True):
+        """Inference the model to compute probabilities.
+
+        Args:
+            prompts (List[List[int]]): List of list of token ids.
```
Comment (lvhan028) [inline lmdeploy/pytorch/decode.py:224]: mismatched docstring

File: tests/pytorch/test_decode.py [hunk 2]
```diff
@@ -0,0 +1,60 @@
...
+from lmdeploy.pytorch.model import init_model
+
+
+def _test_decode_dist(model_path, prompt):
+    tokenizer = AutoTokenizer.from_pretrained(model_path)
+    tokenizer.pad_token_id = tokenizer.eos_token_id
+    tokenizer.padding_side = 'right'
+
+    engine = Engine(model_path, tokenizer=tokenizer)
+    probs = engine.decode(prompt, sort=True, max_bs=4, pad=True)
```
Comment (lvhan028) [inline tests/pytorch/test_decode.py:14]: `prompt` should be tokenized before parsing to `engine.decode`, should it?

File: benchmark/benchmark_decode.py [hunk 3]
```diff
@@ -0,0 +1,90 @@
...
+    start = time.monotonic()
+    # Init an engine
+    engine = Engine(model_path, tokenizer=tokenizer, accel=accel)
+    # decode prompts
+    probs = engine.decode(input_ids)
+    total_tokens = sum(map(len, input_ids))
+
+    elapsed = time.monotonic() - start
+    print(f'Decoded {total_tokens} tokens in {elapsed:.1f} seconds, '
+          f'{total_tokens / elapsed} tokens/s.')
```
Comment (lvhan028) [inline benchmark/benchmark_decode.py:54]: How about  keeping only two decimal places for `token/s`?

File: lmdeploy/pytorch/decode.py [hunk 4]
```diff
@@ -0,0 +1,423 @@
...
+    model_path = args.model_path
+    test_path = args.test_path
+    prompts = args.prompts
+
+    logger = logging.getLogger(__name__)
+    # logging.basicConfig(level=logging.DEBUG)
+
+    # Use test file preferentially
+    if test_path:
+        with open(test_path, 'rb') as f:
```
Comment (lvhan028) [inline lmdeploy/pytorch/decode.py:376]: ”rb“ -> "r"

**Review by lvhan028 — APPROVED** [review]

**Pull request merged**